R/as_model_out_df.R

```diff
@@ -0,0 +1,211 @@
+#' @param output_type_col character string. If an `output_type` column does not
+#' already exist in `tbl`, the `tbl` column name containing `output_type` data.
+#' @param output_type_id_col character string. If an `output_type_id` column does not
+#' already exist in `tbl`, the `tbl` column name containing `output_type_id` data.
+#' @param value_col character string. If an `value` column does not
+#' already exist in `tbl`, the `tbl` column name containing `value` data.
+#' @param sep character string. Character used as separator when concatenating
+#' `team_abbr` and `model_abbr` column values into a single `model_id` string. Only
+#' applicable if `model_id` column not present and `team_abbr` and `model_abbr` columns are.
+#' @param trim_to_task_ids logical. Whether to trim `tbl` to task ID columns only. Task ID
+#' columns can be specified by providing a `<hub_connection>` class object to `hub_con` or
+#' manually through `task_id_cols`.
+#' @param hub_con a `<hub_connection>` class object. Only used if
+#' `trim_to_task_ids = TRUE` and tasks IDs should be determined from the hub config.
+#' @param task_id_cols a character vector of column names. Only used if
+#' `trim_to_task_ids = TRUE` to manually specify task ID columns to retain.
+#' Overrides `hub_con` argument if provided.
+#' @param remove_empty Logical. Whether to remove columns containing only `NA`.
+#'
+#' @return A `model_out_tbl` class object.
+#' @export
+#'
+#' @examples
+#' library(dplyr)
+#' hub_path <- system.file("testhubs/flusight", package = "hubUtils")
+#' hub_con <- connect_hub(hub_path)
+#' hub_con %>%
+#'   filter(output_type == "quantile", location == "US") %>%
+#'   collect() %>%
+#'   filter(forecast_date == max(forecast_date)) %>%
+#'   as_model_out_tbl()
+#'
+as_model_out_tbl <- function(tbl, model_id_col = NULL, output_type_col = NULL,
+                             output_type_id_col = NULL, value_col = NULL, sep = "-",
+                             trim_to_task_ids = FALSE, hub_con = NULL,
+                             task_id_cols = NULL, remove_empty = FALSE) {
+  checkmate::assert_data_frame(tbl)
+  tbl <- tibble::as_tibble(tbl)
+
+  tbl <- rename_columns(tbl,
+    model_id_col = model_id_col,
+    output_type_col = output_type_col,
+    output_type_id_col = output_type_id_col,
+    value_col = value_col
+  )
+
+  if (!"model_id" %in% names(tbl)) {
+    cli::cli_alert_warning("{.arg model_id} column missing. Attempting to create automatically.")
+    tbl <- model_id_merge(tbl, sep = sep)
+  }
+
+  if (trim_to_task_ids) {
+    tbl <- trim_tbl_to_task_ids(
+      tbl = tbl,
+      task_id_cols = task_id_cols,
+      hub_con = hub_con
+    )
+  }
+
+  if (remove_empty) {
+    tbl <- remove_model_out_tbl_empty(tbl)
+  }
+
+  tbl <- std_col_order_model_out_tbl(tbl) %>%
+    std_row_order_model_out_tbl()
+
+  validate_model_out_tbl(tbl)
+
+  structure(tbl,
+    class = c("model_out_tbl", class(tbl))
```

Comment:
I may be missing it, but I don't see a clear statement/definition of what this class is. Is this the only place where we assert that this is a class? Is there a more standard way of defining a S3 class?

Reply:
I guess, after now [consulting with Hadley](https://adv-r.hadley.nz/s3.html#s3-classes), I see that there is no real formal way of defining classes, but he does recommend a validator function. is that something we could/should include in this PR, at least in an initial simple way? This could lead us down a dark rabbit hole of hubValidations too, but it seems that some simple validation should be done when running the as_... function too.

Reply:
Noting that a `validate_model_out_tbl` method is defined

Reply:
whoops, I see that now. should we consider documenting and exporting it?

Reply:
Good call @nickreich . I didn't envisage it as a user facing function, which is why I had not exported it. But if we want to be able to use it in other hubverse packages, it does indeed need to be exported. Done in [7af9e90](https://github.com/Infectious-Disease-Modeling-Hubs/hubUtils/pull/83/commits/7af9e909c3f8f96478b4bef55aec5a13ed94740c)